backend/fms_core/viewsets/_fetch_data.py

```diff
@@ -0,0 +1,133 @@
+import json
+from typing import List
+from django.db.models import OuterRef, Subquery, Exists, Count, Prefetch
+
+from fms.settings import REST_FRAMEWORK
+from fms_core.models import Sample, DerivedBySample, DerivedSample, SampleLineage
+
+def fetch_sample_data(ids: List[int] =[], queryset=None, query_params=None) -> List:
+    """
+    Function used to replace the sample serializer across various viewsets.
+
+    Args:
+        ids: List of ids to select specific samples. Defaults to [].
+        queryset: When called from sample viewset, you can provide the viewset prefiltered queryset. Defaults to None.
+        query_params: You can provide the query params for additional data formatting. Defaults to None.
+
+    Returns:
+        Returns a list of serialized data dictionary (samples)
+    """
+    if queryset is None:
+        derived_samples = DerivedSample.objects.all().select_related('biosample', 'biosample__individual')
+        queryset = Sample.objects.all().prefetch_related(Prefetch('derived_samples', queryset=derived_samples))
+
+    if query_params is not None:
+        limit = int(query_params.get('limit', REST_FRAMEWORK["PAGE_SIZE"]))
+        offset = int(query_params.get('offset', 0))
+
+    # No filtering for empty ids list otherwise use most effective selector
+    if len(ids) == 1:
+        queryset.filter(id=ids[0])
+    elif len(ids) > 1:
+        queryset.filter(id__in=ids)
+
+    samples_queryset = queryset.annotate(
+        first_derived_sample=Subquery(
+            DerivedBySample.objects
+            .filter(sample=OuterRef("pk"))
+            .values_list("derived_sample", flat=True)[:1]
+        )
+    )
+
+    sample_values_queryset = (
+        samples_queryset
+        .annotate(is_library=Exists(samples_queryset.filter(derived_samples__library__isnull=False)))
+        .annotate(derived_count=Count("derived_samples"))
+        .annotate(extracted_from=Subquery(
+                SampleLineage.objects
+                .filter(child=OuterRef("pk"), process_measurement__process__protocol__name="Extraction")
+                .values_list("parent", flat=True)[:1]
+            )
+        )[offset:offset+limit]
+        .values(
+            'id',
+            'name',
+            'container_id',
+            'coordinates',
+            'volume',
+            'concentration',
+            'creation_date',
+            'quality_flag',
+            'quantity_flag',
+            'depleted',
+            'comment',
+            'is_library',
+            'derived_count',
+            'first_derived_sample',
+            'derived_samples__project',
+            'child_of',
+            'extracted_from',
+            'process_measurement',
```

Comment:
for process_measurements you would need .values_list() and add it manually to the sample's query set (the frontend expects a list)